train.py

```diff
@@ -170,10 +162,35 @@
         if k.startswith(unwanted_prefix):
             state_dict[k[len(unwanted_prefix) :]] = state_dict.pop(k)
     model.load_state_dict(state_dict)
-    iter_num = checkpoint["iter_num"]
-    best_val_loss = checkpoint["best_val_loss"]
+    if init_from == 'resume':
+        iter_num = checkpoint["iter_num"]
+        best_val_loss = checkpoint["best_val_loss"]
+
+if init_from == "lora_finetune":
+    out_dir = out_dir + "_lora_finetune"
+    os.makedirs(out_dir, exist_ok=True)
+    for p in model.parameters():
+        p.requires_grad = False
+    apply_lora(model, layer_types=[nn.Linear], dropout=dropout)
+    # Optionally tie the LoRA weights between the embedding layer and the language model output.
+    # I found this didn't help much for Shakespeare, but it might help for other tasks.
+    # tie_lora_weights(model.output, model.tok_embeddings)
+
+    # replace our Task dataset class with the Tiny Shakespeare one.
```

Comment:
not looking good here. hardcoding?

Reply:
Whoops, yeah that's janky. I'll make it configurable. 

Re: looking like multiple PRs, are you referring to the addition of the tinyshakespeare dataset + LoRA? I'd be happy to separate these into two PRs

Reply:
I removed all the tinyshakespeare code from this branch + PR and submitted it as a separate PR (#211). This one can merge on its own, so folks can try fine tuning the pretrained Llama2s on tinystories, or we can tune the tinystories models on Shakespeare if both get merged.